src/util/bson/bson.hpp

```diff
@@ -0,0 +1,190 @@
+/*************************************************************************
+ *
+ * Copyright 2020 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either expreout or implied.
+ * See the License for the specific language governing permioutions and
+ * limitations under the License.
+ *
+ **************************************************************************/
+
+#ifndef REALM_BSON_HPP
+#define REALM_BSON_HPP
+
+#include "indexed_map.hpp"
+#include "regular_expression.hpp"
+#include "min_key.hpp"
+#include "max_key.hpp"
+#include "null.hpp"
+
+#include <realm/binary_data.hpp>
+#include <realm/timestamp.hpp>
+#include <realm/decimal128.hpp>
+#include <realm/object_id.hpp>
+#include <ostream>
+#include <unordered_map>
+#include <variant>
+
+namespace realm {
+namespace bson {
+
+class Bson;
+
+/// A variant of the allowed Bson types.
+class Bson : public std::variant<
+    Null,
+    int32_t,
+    int64_t,
+    bool,
+    float,
+    double,
+    time_t,
+    std::string,
+    std::vector<char>,
+    Timestamp,
+    Decimal128,
+    ObjectId,
+    RegularExpression,
+    MinKey,
+    MaxKey,
+    IndexedMap<Bson>,
+    std::vector<Bson>
+> {
+public:
+    using base = std::variant<
+        Null,
+        int32_t,
+        int64_t,
+        bool,
+        float,
+        double,
+        time_t,
+        std::string,
+        std::vector<char>,
+        Timestamp,
+        Decimal128,
+        ObjectId,
+        RegularExpression,
+        MinKey,
+        MaxKey,
+        IndexedMap<Bson>,
+        std::vector<Bson>
+    >;
+    using base::base;
+    using base::operator=;
+};
+
+using BsonDocument = IndexedMap<Bson>;
+using BsonArray = std::vector<Bson>;
+
+template<class... Ts> struct overloaded : Ts... { using Ts::operator()...; };
+template<class... Ts> overloaded(Ts...) -> overloaded<Ts...>;
+
+inline std::ostream& operator<<(std::ostream& out, const Bson& b)
+{
+    std::visit(overloaded {
+        [&out](auto) {
+            out << "null";
+        },
+        [&out](Null) {
+            out << "null";
+        },
+        [&out](int32_t bson) {
+            out << "{" << "\"$numberInt\"" << ":" << '"' << bson << '"' << "}";
+        },
+        [&out](double bson) {
+            out << "{" << "\"$numberDouble\"" << ":" << '"';
+            if (isnan(bson)) {
```

Comment:
Be explicit with the namespace when using a std function so that the linux builds are happy:
```
../src/util/bson/bson.hpp: In lambda function:
../src/util/bson/bson.hpp:105:17: error: 'isnan' was not declared in this scope; did you mean 'std::isnan'?
  105 |             if (isnan(bson)) {
      |                 ^~~~~
      |                 std::isnan
```